File: backend/models/timetable_model.py
from sqlalchemy import Column, Integer, String, ForeignKey, Time
from sqlalchemy.orm import relationship

from core.database import Base


class TimetableEntryModel(Base):
    """A single class/lecture entry in a student's weekly schedule."""

    __tablename__ = "timetable_entries"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    subject = Column(String, nullable=False)
    instructor = Column(String, nullable=True)
    location = Column(String, nullable=True)
    day_of_week = Column(String, nullable=False)           # Mon, Tue, Wed, Thu, Fri
    start_time = Column(String, nullable=False)            # "09:00"
    end_time = Column(String, nullable=False)              # "10:30"
    color_hex = Column(String, default="#0D6E53")          # Card accent color
    semester = Column(String, nullable=True)               # e.g. "Fall 2024"

File: frontend/lib/features/timetable/data/models/timetable_model.dart
class TimetableEntryModel {
  final int id;
  final int userId;
  final String subject;
  final String? instructor;
  final String? location;
  final String dayOfWeek;
  final String startTime;
  final String endTime;
  final String colorHex;
  final String? semester;

  TimetableEntryModel({
    required this.id,
    required this.userId,
    required this.subject,
    this.instructor,
    this.location,
    required this.dayOfWeek,
    required this.startTime,
    required this.endTime,
    required this.colorHex,
    this.semester,
  });

  factory TimetableEntryModel.fromJson(Map<String, dynamic> json) {
    return TimetableEntryModel(
      id: json['id'],
      userId: json['user_id'],
      subject: json['subject'],
      instructor: json['instructor'],
      location: json['location'],
      dayOfWeek: json['day_of_week'],
      startTime: json['start_time'],
      endTime: json['end_time'],
      colorHex: json['color_hex'] ?? '#0D6E53',
      semester: json['semester'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'subject': subject,
      'instructor': instructor,
      'location': location,
      'day_of_week': dayOfWeek,
      'start_time': startTime,
      'end_time': endTime,
      'color_hex': colorHex,
      'semester': semester,
    };
  }
}
